make the log facility compatible with 64-bit builds

npitre opened the pull request:
To make things clearer, log arguments now have their own type to separate
them from the sea of u32_t, and to allow for a different size on 64-bit
builds. And log records must account for 64-bit pointers, etc.

npitre committed:
log facility: make its arguments 64-bit compatible

Log arguments were hardcoded to u32_t values. On 64-bit systems, this
is rather restrictive. To make things clear, arguments now have their
own type, log_arg_t, which now can be adjusted in only one location
if need be.  It is currently defined as unsigned long whose effective
width is equivalent to u32_t on 32-bit systems, and u64_t on 64-bit
systems.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre committed:
log facility: make its records 64-bit compatible

Log records may store either data or pointers to more records. In both
cases they must have the same size. With 64-bit pointers, the amount
of data that can occupy the same space as a pointer has to be adjusted.
And storage alignment has to accommodate actual pointers not u32_t.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre commented:
This requires PR #16586 to be applied for the Kconfig issue to be resolved.

nordic-krch commented on include/logging/log_msg.h:
```diff
@@ -22,22 +22,33 @@ extern "C" {
 
 /** @brief Number of arguments in the log entry which fits in one chunk.*/
+#ifdef CONFIG_64BIT
+#define LOG_MSG_NARGS_SINGLE_CHUNK 4
```
what was the reason for that change? to have single chunk 64 bytes long?

nordic-krch commented on tests/subsys/logging/log_msg/src/log_msg_test.c:
```diff
@@ -20,11 +20,16 @@ extern struct k_mem_slab log_msg_pool;
 static const char my_string[] = "test_string";
 void test_log_std_msg(void)
```
maybe you could rework this test a bit and create a loop that tests up to six arguments and determines expected number of chunks allocated like `(arg > LOG_MSG_NARGS_SINGLE_CHUNK) ? 2 : 1`. That would look better and would not need ifdef.

nordic-krch commented on tests/subsys/logging/log_msg/src/log_msg_test.c:
```diff
@@ -20,11 +20,16 @@ extern struct k_mem_slab log_msg_pool;
 static const char my_string[] = "test_string";
 void test_log_std_msg(void)
 {
+#ifdef CONFIG_64BIT
```
what about something like:
```
int exp_nargs_in_single_chunk = IS_ENABLED(CONFIG_64BIT) ? 4 : 3;
zassert_true(LOG_MSG_NARGS_SINGLE_CHUNK == exp_nargs_in_single_chunk , "test assumes following setting");
```

nordic-krch reviewed (changes requested):
Look ok (minor comments). Does string formatting works with that or does it need additional changes?

npitre committed:
log_msg test: rework test_log_std_msg() to better cope with 64-bit builds

Remove #ifdef's and use a loop to avoid duplicating the code.

Suggested by Krzysztof Chruscinski.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre commented:
On Sun, 9 Jun 2019, Krzysztof Chruscinski wrote:

> Look ok (minor comments). Does string formatting works with that or does it need additional changes?

String formatting requires its own set of changes.

>  /** @brief Number of arguments in the log entry which fits in one chunk.*/
> +#ifdef CONFIG_64BIT
> +#define LOG_MSG_NARGS_SINGLE_CHUNK 4
> 
> what was the reason for that change? to have single chunk 64 bytes long?

This determines the size of union log_msg_head_data which has to match 
the size of struct log_msg_ext_head_data not to waste memory as they 
occupy the same payload space in struct log_msg. There is a build-time 
assertion enforcing that. And the size of struct log_msg_ext_head_data 
gets bigger due to the pointer and the argument type it contains which 
are larger on 64-bit targets.

> > @@ -20,11 +20,16 @@ extern struct k_mem_slab log_msg_pool;
>  static const char my_string[] = "test_string";
>  void test_log_std_msg(void)
> 
> maybe you could rework this test a bit and create a loop that tests up 
> to six arguments and determines expected number of chunks allocated 
> like `(arg > LOG_MSG_NARGS_SINGLE_CHUNK) ? 2 : 1`. That would look 
> better and would not need ifdef.

Indeed, I like your suggestion.

I'm adding an extra patch to this PR implementing those changes 
separately, in case they have unexpected side effects.

aescolar commented:
@npitre: Note that this PR is blocked due to checkpatch errors.

npitre commented:
On Thu, 20 Jun 2019, Alberto Escolar wrote:

> Note that this PR is blocked due to checkpatch errors.

I'd argue that checkpatch is wrong in this particular context.
Can it be bypassed?

nashif commented:
> I'd argue that checkpatch is wrong in this particular context. Can it be bypassed?

they are not wrong, we following the linux coding style and you need to have the statement and the break on a new line in this case.

npitre commented:
On Thu, 20 Jun 2019, Anas Nashif wrote:

> they are not wrong, we following the linux coding style and you need to have the statement and the break on a new line in this case.

The Linux community uses a coding style and checkpatch as guidelines, 
not as a hard requirement for patch merging. Human common sense always 
prevails over checkpatch when readability of the code is improved in 
some circumstances. Linus Torvalds himself made public statements on a 
few occasions against unconditional checkpatch enforcement as it 
sometimes leads to worse code.
